mrjob/queue.py

```diff
@@ -0,0 +1,119 @@
+# -*- coding: utf-8 -*-
+# Copyright 2009-2019 Yelp and Contributors
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+# http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+import datetime
+import logging
+import os.path
+import time
+
+
+log = logging.getLogger(__name__)
+
+
+class SchedulingQueue(object):
+    """This class is used when there are no running clusters in the desired
+    pool. In this situation we wish to spin up a new cluster, but we do not
+    want to spin up one for every process currently running. Therefore this
+    class defines a light-weight scheduling queue. This queue is built in S3
+    to prevent introducing any new infrastructure and only performs writes of
+    new objects and list operations to achieve read-after-write semantics."""
+
+    def __init__(self, s3_fs, s3_prefix, queue_length, runner_self):
+        """
+        :param s3_fs: File system to use when interacting with S3.
+        :param s3_prefix: Where we store store all of our markers.
+                          This must be unique to the target pool.
+        :param queue_length: The maximum number of clusters that we allow
```

Comment:
I would call this max_allowed_cluster, since its not really the length of the queue but rather how many items you are allowed to remove from it.